BUIP055

gandrewstone opened the pull request:
BUIP055 implementation, including tests, fork configuration, and handling OP_RETURN fork-invalid transactions.

ptschip committed:
Handle headers that won't connect right away.

At times we can receive out of order headers.  If we do we store
the unconnecting header in a cache and check if it will connect
each time a new header arrives.

Out of order headers should typically
exist for only a few seconds at most, therefore if the unconnecting
header is in cache for more than 2 minutes we will delete it.

There is also the possiblity of an attack to prevent node sync from
happening during the intial download of headers.  If we get out of
order headers and we are far from today then we disconnect the peer
so that we can start downloading initial headers from another peer.

ptschip committed:
Add more tests for unconnecting headers in sendheaders.py

gandrewstone committed:
BUIP055 implementation including fork control and invalid OP_RETURN transaction, and relevant tests.

Commit:
Don't send connection maintenance messages until handshake complete

- get rid of unnecessary rejects list, send block rejects immediately
- use the correct test for early return in SendMessages()
- break out the ban check to the connection manager, and do it
  before this test so that fDisconnect is set appropriately

sickpig committed:
Avoid using floating point arithmetic to compute EB

Change `excessiveBlockSize` type to uint64_t since has to be
coherent with `maxGeneratedBlock`. Expand also the tests for
EB signaling in subVersion string.

gandrewstone committed:
Both CTweak and CTweakRef were not handled by allowed args.  Fix compilation error in debug mode

gandrewstone committed:
formatting

sickpig committed:
uint64_t type for aux vars containing EB and MGS

sickpig committed:
Use uint64_t for EB and MGS also in `unlimitedmodel.cpp`

ptschip committed:
Also check pindexBestHeader when determining fScriptChecks

Since we can be reasonable sure we have all the headers when we
do IBD or receive new blocks then we can also check if we are
within approx 30 days of the chain tip by block height.  We still
keep the check that timeBarrier is also within 30 days but by
having both checks it makes it more difficult to attack in some
way.

sickpig committed:
Use `qulonglong` rather than uint64_t per http://doc.qt.io/qt-5/qvariant.html

deadalnix commented on src/main.cpp:
```diff
@@ -4108,6 +4118,15 @@ bool ContextualCheckBlockHeader(const CBlockHeader &block, CValidationState &sta
+        if (block.GetAncestor(forkActivationHeight) != forkBlockHash)
+            return state.Invalid(error("%s : rejected wrong fork block", __func__), REJECT_OBSOLETE, "wrong-fork");
+    }
+#endif     
```
Remove

deadalnix commented on src/chain.h:
```diff
@@ -218,6 +223,18 @@ class CBlockIndex
+    bool forkActivateNow(int time);
+
+    // Is the next block the fork block?
+    bool forkAtNextBlock(int time);
```
These name would benefit from being verbs rather than nouns.

deadalnix commented on src/main.cpp:
```diff
@@ -4156,9 +4175,31 @@ bool ContextualCheckBlock(const CBlock &block, CValidationState &state, CBlockIn
+            return state.DoS(100,
+                error("%s: BUIP055 fork block (%s, height %d) must exceed %d, but this block is %d bytes", __func__,
+                                 hash.ToString(), nHeight, BLOCKSTREAM_CORE_MAX_BLOCK_SIZE, block.nBlockSize),
+                REJECT_INVALID, "bad-fork-block");
```
Use bad-blk-too-small so we can have the same error codes.

deadalnix commented on src/main.cpp:
```diff
@@ -4156,9 +4175,31 @@ bool ContextualCheckBlock(const CBlock &block, CValidationState &state, CBlockIn
 }
 
+
+
```
remove

deadalnix commented on src/test/fork_tests.cpp:
```diff
@@ -0,0 +1,57 @@
+#include "net.h"
+#include "chainparams.h"
+
+using namespace std;
```
You shouldn't blanket import namespaces.

gandrewstone replied:
removed

deadalnix commented on src/globals.cpp:
```diff
@@ -178,6 +178,12 @@ CTweakRef<uint64_t> miningBlockSize("mining.blockSize",
+
+CTweak<uint64_t> miningForkTime("mining.forkTime",
+    "Time in seconds since the epoch to initiate a hard fork as per BUIP055.",
+    1501545600); // Tue 1 Aug 2017 00:00:00 UTC
```
It's actually not midnight, but 8:20pm or so if memory serves.

deadalnix commented on src/buip055fork.h:
```diff
@@ -0,0 +1,11 @@
+#ifndef BUIP055FORK_H
```
Copyright header is missing.

gandrewstone replied:
added

deadalnix reviewed (commented):
Nothing super fancy, but a few things can be improved.

gandrewstone committed:
Merge pull request #681 from gandrewstone/allowedargs_ctweaks

Both CTweak and CTweakRef were not handled by allowed args.

gandrewstone committed:
Merge pull request #687 from ptschip/dev_checkpoints

Also check pindexBestHeader when determining fScriptChecks

ptschip committed:
Add 4 block attack scenarios to PV python testing

Also add the needed option -pvtest which slows down the checking
of inputs by putting a sleep time of 1 second for each input to
check.  This allows up to simulate very long to validate blocks
used in the attack scenarios.

gandrewstone committed:
match against altnames and the CN (common name) so that different spam and hosting solutions work.  Fix the short read error (fix provided by @bitcartel).

gandrewstone committed:
auto-format

gandrewstone committed:
Merge pull request #682 from sickpig/new/improve-eb-in-mb-computation

Avoid using floating point arithmetic to compute EB

gandrewstone committed:
Merge pull request #693 from gandrewstone/bitnodes_cert

Bitnodes match against altnames so anti-spam and other hosting works.  Fix short read error.

gandrewstone committed:
Merge pull request #652 from kyuupichan/msg_processing

Improve message processing so we don't send pings, etc., before handshake is complete

gandrewstone committed:
Merge pull request #650 from ptschip/dev_bans2

Handle non continuous headers

ptschip committed:
Check for reorgs in progress when new blocks are beginning validation.

Remove unnecessary code

gandrewstone committed:
fix rare test error

sickpig commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -150,28 +151,34 @@ def hex_str_to_bytes(hex_str):
     while True:
         pool = set(rpc_connections[0].getrawmempool())
         num_match = 1
+        poolLen = [len(pool)]
```
nit pick: why using camelcase here?

sickpig commented on src/main.h:
```diff
@@ -113,7 +113,7 @@ static const int64_t BLOCK_DOWNLOAD_TIMEOUT_BASE = 1000000;
 static const int64_t BLOCK_DOWNLOAD_TIMEOUT_PER_PEER = 500000;
 /** Timeout in secs for the initial sync. If we don't receive the first batch of headers */
-static const uint32_t INITIAL_HEADERS_TIMEOUT = 30;
+static const uint32_t INITIAL_HEADERS_TIMEOUT = 120;
```
why is this change needed

sickpig commented on src/buip055fork.cpp:
```diff
@@ -0,0 +1,115 @@
+#include "utilstrencodings.h"
+#include "validationinterface.h"
+#include "version.h"
+
```
do we really need all these includes?

sickpig commented on src/buip055fork.cpp:
```diff
@@ -0,0 +1,115 @@
+bool ValidateBUIP055Tx(const CTransaction& tx)
+{
+    bool ret = !IsTxOpReturnInvalid(tx);
+    // TODO: validate new signature (REQ-6-2, REQ-6-3)
```
REQ-2/3 = replay protection from big fork txns being included in the other chain using the new hash scheme i.e. #4. So that mean that we still are going to complete this check once #4 will be merged, right?

sickpig commented on src/buip055fork.cpp:
```diff
@@ -0,0 +1,115 @@
+    for (const CTransaction &tx : block.vtx)
+    {
+        if (!ValidateBUIP055Tx(tx))
+            return state.DoS(100,
```
setting DoS to 100 mean that we are going to ban the peer who relayed to us the txn even if it is not the one who generated the txn in the first place, is this something that we really wanted to?

sickpig commented on src/buip055fork.cpp:
```diff
@@ -0,0 +1,115 @@
+    return false;
+}
+
+bool ValidateBUIP055Block(const CBlock &block, CValidationState &state)
```
I cant's see anywhere the code that implement the sunset rule that consider this kind of txn invalid only if block height <= 530000, see REQ-6.1.

ptschip committed:
Move block relay to ActivateBestChainStep

In PV we can connect blocks beyone pindexMostWork if other
blocks arrive just after we start validation.  Therefore we
need to relay block inventory based on the new pindexMostWork
after all the blocks have been connected.  If we instead
return to ActivateBestChain we will relay inventory based on the
old pindexMostWork and may end up not relaying a block.

ptschip committed:
Indent section

No changes to logic, just indentation and comment changes.

ptschip committed:
Add extended test for multiple block validation

Mine  blocks from different nodes at the same time repeatedly
for 100 iterations.

ptschip committed:
Add extended reorg tests for PV

There are many different reorg possibilities that are tested
in this addition to the python tests.

gandrewstone committed:
Merge branch 'dev' into buip-hf

ptschip committed:
Fix edge case with multi reorgs

Handle a reorg that is not yet complete but another reorg
is initiated.

ptschip committed:
Run clang-format on main.cpp

gandrewstone committed:
fix excessive.py extended tests to work faster and with the xthin excessive size changes

ptschip committed:
return if pindexBestHeader is NULL

FindMostWorkChain() can return an NULL.

Also updated fix for compiler warning. Casting pindex->nHeight to
uint32_t.  TODO: why is pindex->nHeight an int, it really should
be uint anway.

Add Parenthesis for code clarity.

gandrewstone committed:
Merge pull request #696 from ptschip/dev_pv

PV: reorg bug fix (issue #566)

gandrewstone committed:
Merge pull request #704 from gandrewstone/qa_largeTxns

fix excessive.py extended tests to work faster and with the xthin exc…

sickpig committed:
Add sunset clause to REQ-6-1 implementation.

As per UAHF spec (*) section REQ-6-1 once the fork has activated,
transactions containing an OP_RETURN output with a specific magic data
value of

`Bitcoin: A Peer-to-Peer Electronic Cash System`

(46 characters, including the single spaces separating the words, and
without any terminating null character) shall be considered invalid
until block 530,000 inclusive.

This commit just add the sunset clause, i.e. block height lower than
530,001.

(*) https://github.com/Bitcoin-UAHF/spec/blob/master/uahf-technical-spec.md#req-6-1-disallow-special-op_return-marked-transactions-with-sunset-clause

sickpig committed:
Update the REQ-6-1 anti reply magic value

While at it use a string to define it. That way
is more clear and more to the point we do the
same also in the python tests.

gandrewstone committed:
Merge pull request #2 from sickpig/add-req-6-1-sunset-clause

Add sunset clause to REQ-6-1 implementation.

gandrewstone committed:
Merge pull request #3 from sickpig/update-anti-reply-magic-value

Update anti reply magic value

gandrewstone committed:
merge
